### PR TITLE
MULE-10927 - Support http policy specific pointcuts - adding support …

### PR DESCRIPTION
**Pull request by pablolagreca**
…for policies pointcuts for path and method in request and listener

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/policy/DefaultPolicyManager.java [hunk 1]
```diff
@@ -38,27 +40,33 @@
 
   private Collection<OperationPolicyParametersTransformer> operationPolicyParametersTransformerCollection = emptyList();
   private Collection<SourcePolicyParametersTransformer> sourcePolicyParametersTransformerCollection = emptyList();
+  private Collection<SourcePolicyPointcutParametersFactory> sourcePointcutFactories = emptyList();
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/policy/DefaultPolicyManager.java]: PolicyManager does not look cohesive. There are two sets of methods that are used in different contexts, one for sources, and another one for operations. 
The relation between the findXPolicyInstance and the createXPointcutParameters is strange too, as they will be always used together, so why not merge both methods in one?

File: extensions/http/src/test/java/org/mule/extension/http/policy/HttpListenerPolicyPointcutParametersFactoryTestCase.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+public class HttpListenerPolicyPointcutParametersFactoryTestCase extends AbstractMuleTestCase {
+
+  private static final ComponentIdentifier HTTP_LISTENER_COMPONENT_IDENTIFIER =
+      new ComponentIdentifier.Builder().withNamespace("http").withName("listener").build();
+  private static final String TEST_REQUEST_PATH = "test-request-path";
+  private static final String TEST_METHOD = "PUT";
+
+  private final HttpListenerPolicyPointcutParametersFactory factory = new HttpListenerPolicyPointcutParametersFactory();
+  private final HttpRequestAttributes httpAttributes = Mockito.mock(HttpRequestAttributes.class);
```
Comment (pablokraan) [inline extensions/http/src/test/java/org/mule/extension/http/policy/HttpListenerPolicyPointcutParametersFactoryTestCase.java]: static imports

File: extensions/http/src/test/java/org/mule/extension/http/policy/HttpListenerPolicyPointcutParametersFactoryTestCase.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  public void supportsHttpListener() {
+    assertThat(factory
+        .supportsSourceIdentifier(HTTP_LISTENER_COMPONENT_IDENTIFIER),
+               is(true));;
+  }
+
+  @Test
+  public void doesNotSupportHttpRequester() {
+    assertThat(factory
+        .supportsSourceIdentifier(new ComponentIdentifier.Builder().withNamespace("http").withName("request").build()),
```
Comment (pablokraan) [inline extensions/http/src/test/java/org/mule/extension/http/policy/HttpListenerPolicyPointcutParametersFactoryTestCase.java]: static import

File: extensions/http/src/test/java/org/mule/extension/http/policy/HttpRequestPolicyPointcutParametersFactoryTestCase.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  private static final ComponentIdentifier HTTP_REQUEST_COMPONENT_IDENTIFIER =
+      new ComponentIdentifier.Builder().withNamespace("http").withName("request").build();
+  private static final String TEST_REQUEST_PATH = "test-request-path";
+  private static final String TEST_METHOD = "PUT";
+
+  private final HttpRequestPolicyPointcutParametersFactory factory = new HttpRequestPolicyPointcutParametersFactory();
+
+  @Test
+  public void supportsHttpListener() {
```
Comment (pablokraan) [inline extensions/http/src/test/java/org/mule/extension/http/policy/HttpRequestPolicyPointcutParametersFactoryTestCase.java]: Fix test names

File: core/src/main/java/org/mule/runtime/core/policy/PolicyPointcutParameters.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * This class defines a set of parameters related to the {@link org.mule.runtime.core.api.source.MessageSource} and the message
+ * received by the {@link org.mule.runtime.core.api.source.MessageSource}. These parameters will later by used during policy
+ * lookup in {@link PolicyProvider} to match it against a configured pointcut.
+ * 
+ * @since 4.0
+ */
+public class PolicyPointcutParameters {
+
+  private ComponentIdentifier componentIdentifier;
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/policy/PolicyPointcutParameters.java]: final

File: core/src/main/java/org/mule/runtime/core/policy/PolicyPointcutParameters.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+   */
+  public PolicyPointcutParameters(ComponentIdentifier componentIdentifier) {
+    this.componentIdentifier = componentIdentifier;
+  }
+
+  /**
+   * @return the component identifier. This is the namespace of the module were it is defined and the source / operation
+   *         identifier.
+   */
+  public ComponentIdentifier getComponentIdentifier() {
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/policy/PolicyPointcutParameters.java]: final?

File: modules/http/src/main/java/org/mule/extension/http/api/policy/HttpRequestPolicyPointcutParametersFactory.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+package org.mule.extension.http.api.policy;
+
+import org.mule.runtime.core.policy.OperationPolicyPointcutParametersFactory;
+import org.mule.runtime.core.policy.PolicyPointcutParameters;
+import org.mule.runtime.dsl.api.component.ComponentIdentifier;
+
+import java.util.Map;
+
+/**
+ * Http request operation policy pointcut parameters factory.
```
Comment (pablokraan) [inline modules/http/src/main/java/org/mule/extension/http/api/policy/HttpRequestPolicyPointcutParametersFactory.java]: Http > HTTP

File: modules/http/src/main/java/org/mule/extension/http/api/policy/HttpListenerPolicyPointcutParametersFactory.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+import static org.mule.runtime.api.util.Preconditions.checkArgument;
+import org.mule.extension.http.api.HttpRequestAttributes;
+import org.mule.runtime.api.message.Attributes;
+import org.mule.runtime.core.policy.PolicyPointcutParameters;
+import org.mule.runtime.core.policy.SourcePolicyPointcutParametersFactory;
+import org.mule.runtime.dsl.api.component.ComponentIdentifier;
+
+/**
+ * Http request operation policy pointcut parameters factory.
```
Comment (pablokraan) [inline modules/http/src/main/java/org/mule/extension/http/api/policy/HttpListenerPolicyPointcutParametersFactory.java]: Http > HTTP

File: tests/functional/src/main/java/org/mule/functional/util/http/SimpleHttpServer.java [hunk 9]
```diff
@@ -75,7 +75,7 @@ public void stop() {
 
     @Override
     public void handle(HttpExchange exchange) throws IOException {
-      httpRequests.push(new HttpMessage(exchange));
+      httpRequests.addLast(new HttpMessage(exchange));
```
Comment (pablokraan) [inline tests/functional/src/main/java/org/mule/functional/util/http/SimpleHttpServer.java]: Why is this change required?

File: modules/http/src/main/java/org/mule/extension/http/api/policy/HttpRequestPolicyPointcutParametersFactory.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+  @Override
+  public boolean supportsOperationIdentifier(ComponentIdentifier operationIdentifier) {
+    return requestIdentifier.equals(operationIdentifier);
+  }
+
+  @Override
+  public PolicyPointcutParameters createPolicyPointcutParameters(ComponentIdentifier operationIdentifier,
+                                                                 Map<String, Object> operationParameters) {
+    String pathParameter = (String) operationParameters.get("path");
```
Comment (pablokraan) [inline modules/http/src/main/java/org/mule/extension/http/api/policy/HttpRequestPolicyPointcutParametersFactory.java]: Defines constants for "path" and "method" and use them on the tests

File: tests/functional/src/main/java/org/mule/functional/util/http/SimpleHttpServer.java (hunk 9 above)
Comment (pablolagreca) [inline tests/functional/src/main/java/org/mule/functional/util/http/SimpleHttpServer.java]: Because it was adding the HTTP messages out of order.

File: core/src/main/java/org/mule/runtime/core/policy/SourcePolicyResult.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+ * Result of executing a {@link SourcePolicy}.
+ *
+ * It contains the {@link Event} result of the flow execution and also allows to generate
+ * the response parameters and error response parameters to be sent by the source.
+ *
+ * @since 4.0
+ */
+public interface SourcePolicyResult {
+
+  Either<MessagingException, Event> getResult();
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/policy/SourcePolicyResult.java]: javadoc?

File: core/src/main/java/org/mule/runtime/core/policy/SuccessSourcePolicyResult.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  private final Map<String, Object> responseParameters;
+  private final MessageSourceResponseParametersProcessor messageSourceResponseParametersProcessor;
+
+  /**
+   * Creates a new successful result from a policy execution.
+   *
+   * @param flowExecutionResult the result of the flow execution.
+   * @param responseParameters the response parameters to be sent by the source.
+   */
+  public SuccessSourcePolicyResult(Event flowExecutionResult, Map<String, Object> responseParameters,
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/policy/SuccessSourcePolicyResult.java]: Add missing param to the javadoc

File: core/src/main/java/org/mule/runtime/core/policy/SourcePolicyPointcutParametersFactory.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+   * @return true if this factory can create {@link PolicyPointcutParameters} for the source identifier, false otherwise.
+   */
+  boolean supportsSourceIdentifier(ComponentIdentifier sourceIdentifier);
+
+  /**
+   * Creates an specific {@link PolicyPointcutParameters} for a particular source operation by {@code sourceIdentifier}.
+   * 
+   * @param sourceIdentifier identifier of the message source
+   * @param attributes the attributes from the message generated by the message source
+   * @return
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/policy/SourcePolicyPointcutParametersFactory.java]: returns?

File: core/src/main/java/org/mule/runtime/core/policy/SourcePolicyProcessor.java [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+  private final PolicyEventConverter policyEventConverter = new PolicyEventConverter();
+  private final Processor nextProcessor;
+
+  /**
+   * Creates a new {@code DefaultSourcePolicy}.
+   * 
+   * @param policy the policy to execute before and after the source.
+   * @param policyStateHandler the state handler for the policy.
+   * @param nextProcessor the next-operation processor implementation, it may be another policy or the flow execution.
+   * @param nextProcessor
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/policy/SourcePolicyProcessor.java]: Remove duplicated param

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/operation/InterceptingOperationMessageProcessorTestCase.java [hunk 15]
```diff
@@ -16,19 +16,21 @@
 import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
-
+import org.mule.runtime.core.api.Event;
```
Comment (pablokraan) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/operation/InterceptingOperationMessageProcessorTestCase.java]: Are these changes required?

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationMessageProcessorTestCase.java [hunk 16]
```diff
@@ -308,17 +308,17 @@ public void executesWithDefaultConfig() throws Exception {
...
-    when(mockPolicyManager.findOperationPolicy(eventContextId, operationIdentifier)).thenReturn(of(mockPolicy));
-    when(extensionModel.getName()).thenReturn(EXTENSION_NAMESPACE);
-    when(operationModel.getName()).thenReturn(OPERATION_NAME);
-
-    messageProcessor.process(event);
-
-    verify(mockPolicyManager).findOperationPolicy(eventContextId, operationIdentifier);
-    verify(mockPolicy.process(same(event), any(Processor.class), any(OperationParametersProcessor.class)));
+    //TODO fix
+    //String eventContextId = event.getContext().getId();
```
Comment (pablokraan) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationMessageProcessorTestCase.java]: Fix this

File: core/src/main/java/org/mule/runtime/core/policy/DefaultPolicyManager.java [hunk 17]
```diff
@@ -38,28 +42,55 @@
...
   @Override
-  public Optional<OperationPolicy> findOperationPolicy(String executionIdentifier, ComponentIdentifier operationIdentifier) {
-    List<Policy> parameterizedPolicies = policyProvider.findOperationParameterizedPolicies(operationIdentifier);
+  public OperationPolicy createOperationPolicy(ComponentIdentifier operationIdentifier, Event event,
+                                               Map<String, Object> operationParameters,
+                                               OperationExecutionFunction operationExecutionFunction) {
+
+    PolicyPointcutParameters operationPointcutParameters =
+        createOperationPointcutParameters(operationIdentifier, operationParameters);
+    List<Policy> parameterizedPolicies = policyProvider.findOperationParameterizedPolicies(operationPointcutParameters);
```
Comment (pablolagreca) [inline core/src/main/java/org/mule/runtime/core/policy/DefaultPolicyManager.java]: I removed findOperationParameterizedPolicies from the API since it's only being used internally. And now it's private

File: core/src/main/java/org/mule/runtime/core/policy/DefaultPolicyManager.java [hunk 18]
```diff
@@ -90,13 +121,37 @@ public void initialise() throws InitialisationException {
...
     } catch (RegistrationException e) {
       throw new InitialisationException(e, this);
     }
   }
 
+  private PolicyPointcutParameters createSourcePointcutParameters(ComponentIdentifier sourceIdentifier, Event sourceEvent) {
+    return sourcePointcutFactories.stream()
+        .filter(sourcePolicyPointcutParametersFactory -> sourcePolicyPointcutParametersFactory
+            .supportsSourceIdentifier(sourceIdentifier))
+        .findAny()
```
Comment (pablolagreca) [inline core/src/main/java/org/mule/runtime/core/policy/DefaultPolicyManager.java]: That shouldn't be the case. I will add a condition to fail if that's the case

File: core/src/main/java/org/mule/runtime/core/policy/DefaultPolicyManager.java [hunk 19]
```diff
@@ -90,13 +121,37 @@ public void initialise() throws InitialisationException {
...
+            .createPolicyPointcutParameters(sourceIdentifier, sourceEvent.getMessage().getAttributes()))
+        .orElse(new PolicyPointcutParameters(sourceIdentifier));
+  }
+
+  private PolicyPointcutParameters createOperationPointcutParameters(ComponentIdentifier operationIdentifier,
+                                                                     Map<String, Object> operationParameters) {
+    return operationPointcutFactories.stream()
+        .filter(sourcePolicyPointcutParametersFactory -> sourcePolicyPointcutParametersFactory
+            .supportsOperationIdentifier(operationIdentifier))
+        .findAny()
```
Comment (pablolagreca) [inline core/src/main/java/org/mule/runtime/core/policy/DefaultPolicyManager.java]: That shouldn't be the case. I will add a condition to fail if that's the case

File: core/src/main/java/org/mule/runtime/core/policy/AbstractCompositePolicy.java [hunk 20]
```diff
@@ -54,39 +55,45 @@ public AbstractCompositePolicy(List<Policy> policies,
    * in the chain until the finally policy it's executed in which case then next operation of it, it will be the operation
    * execution.
    */
-  public Event process(Event operationEvent, Processor nextProcessor, ParametersProcessor parametersProcessor)
-      throws Exception {
-    return new AbstractCompositePolicy.NextOperationCall(operationEvent, nextProcessor, parametersProcessor)
+  public final Event processPolicies(Event operationEvent) throws Exception {
+    return new AbstractCompositePolicy.NextOperationCall(operationEvent)
```
Comment (pablolagreca) [inline core/src/main/java/org/mule/runtime/core/policy/AbstractCompositePolicy.java]: Can't do that since is new of an inner calss.

File: core/src/main/java/org/mule/runtime/core/policy/PolicyPointcutParameters.java [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+import org.mule.runtime.dsl.api.component.ComponentIdentifier;
+
+/**
+ * This class defines a set of parameters related to the {@link org.mule.runtime.core.api.source.MessageSource} and the message
+ * received by the {@link org.mule.runtime.core.api.source.MessageSource}. These parameters will later by used during policy
+ * lookup in {@link PolicyProvider} to match it against a configured pointcut.
+ * 
+ * @since 4.0
+ */
+public class PolicyPointcutParameters {
```
Comment (pablolagreca) [inline core/src/main/java/org/mule/runtime/core/policy/PolicyPointcutParameters.java]: No, it's going to be the default implementation when there's no particular factory for parameters

File: core/src/main/java/org/mule/runtime/core/policy/SourcePolicyResult.java (hunk 11 above)
Comment (pablolagreca) [inline core/src/main/java/org/mule/runtime/core/policy/SourcePolicyResult.java]: removed

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/operation/InterceptingOperationMessageProcessorTestCase.java (hunk 15 above)
Comment (pablolagreca) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/operation/InterceptingOperationMessageProcessorTestCase.java]: yes, since I'm wrapping the execution in the policy I need the policy mock to execute the real operation callback.

Comment (pablokraan) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/operation/InterceptingOperationMessageProcessorTestCase.java]: I don't think that is required, there are changes in the import, but no changes in the code below

File: core/src/main/java/org/mule/runtime/core/policy/SuccessSourcePolicyResult.java [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+   */
+  public Map<String, Object> getResponseParameters() {
+    return responseParameters;
+  }
+
+  /**
+   * @param event the event to use as source for generating the error response parameters.
+   * @return the set of parameters to execute the function that sends the failure response.
+   */
+  public Map<String, Object> createErrorResponseParameters(Event event) {
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/policy/SuccessSourcePolicyResult.java]: I don't understand why is this method required if this class is used for successful results only (there are another class for error results)

Comment (pablolagreca) [inline core/src/main/java/org/mule/runtime/core/policy/SuccessSourcePolicyResult.java]: Because even though this is a successful response it may fail when evaluating the response parameters. If that's the case the exception strategy needs to get executed and the error response parameters must be executed with the result of the exceptions strategy.
